modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CacheExchangeMergeTest.java

```diff
@@ -1598,10 +1614,88 @@ private IgniteInternalFuture startGridsAsync(Ignite node, int startIdx, int cnt)
-                return ((GridDhtPartitionsAbstractMessage)msg).exchangeId() != null || (msg instanceof GridDhtPartitionsSingleRequest);
+            return delay(msg);
+        }
+    }
 
+    /**
+     * Return {@code true} if need to delay message to emulate merge exchanges.
+     *
+     * @param msg Message.
+     * @return {@code True} if need to delay message.
+     */
+    private static boolean delay(Message msg) {
+        if (!GridDhtPartitionsAbstractMessage.class.isInstance(msg))
             return false;
+
+        GridDhtPartitionsAbstractMessage dhtMsg = (GridDhtPartitionsAbstractMessage)msg;
+        return nonNull(dhtMsg.exchangeId()) || GridDhtPartitionsSingleRequest.class.isInstance(dhtMsg);
+    }
+
+    /**
+     * Test checks that there will be no {@link ConcurrentModificationException}
+     * when merging exchanges and iterating over {@link ExchangeDiscoveryEvents#events} at the same time.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testNoConcurrentModificationExceptionAfterMergeExchanges() throws Exception {
+        testSpi = true;
+
+        LogListener logLsnr = matches("Merge exchange future on finish [").build();
+        listeningLog.registerAllListeners(logLsnr);
+
+        startGrid(0);
+
+        for (int i = 1; i < 9; i++) {
+            IgniteConfiguration cfg = getConfiguration(getTestIgniteInstanceName(i));
+            TestRecordingCommunicationSpi spi = ((TestRecordingCommunicationSpi)cfg.getCommunicationSpi());
+
+            spi.blockMessages((node, msg) -> delay(msg));
+            runAsync(() -> startGrid(cfg), "create-node-" + cfg.getIgniteInstanceName());
+            spi.waitForBlocked();
+        }
+
+        List<Ignite> allNodes = IgnitionEx.allGridsx();
+
+        AtomicBoolean stop = new AtomicBoolean();
+        Collection<Exception> exceptions = new ConcurrentLinkedQueue<>();
+        CountDownLatch latch = new CountDownLatch(allNodes.size());
+
+        for (Ignite gridEx : allNodes) {
+            runAsync(() -> {
+                Collection<DiscoveryEvent> evts = ((IgniteEx)gridEx).context().cache().context().exchange()
+                    .lastTopologyFuture().events().events();
+
+                latch.countDown();
+
+                int i = 0;
+                while (!stop.get()) {
+                    try {
+                        for (DiscoveryEvent evt : evts) {
+                            if (nonNull(evt))
+                                i++;
+                        }
+                    }
+                    catch (ConcurrentModificationException e) {
+                        exceptions.add(e);
+
+                        log.error("i = " + i, e);
+
+                        break;
+                    }
+                }
+            }, "get-ex-" + gridEx.configuration().getIgniteInstanceName());
         }
+
+        for (Ignite node : allNodes)
+            TestRecordingCommunicationSpi.spi(node).stopBlock();
+
```

Comment:
Please wrap 3 lines below to finally block for avoiding test hanging, if something goes wrong.